### ISSUE TITLE
Feature Request: Allow mixed box type to contain multiple arbitrary aliquot types

### ISSUE THREAD

**Issue by saipavan10-git** [dr01d3r/redcap-em-orca-specimen-tracking#3]
Hi,

I work with @PBChase @chemikyle at UF. We want to use your module for our own bio-specimen tracking but we are having trouble with the restriction on box-aliquot matching. Specifically the `box_type` has to match with `sample_type`

Ideally, we want to create a mixed box type and have it accept any type of aliquot (including multiple different types of aliquot). A partially populated box would look like this:

|       |       |       |       |       |       |       |       |       |       |
|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|
| BC-01 | BC-02 | BC-03 | PG-01 | PG-02 | PG-03 | PG-04 | PG-05 | PG-06 | PG-07 |
| PG-08 | S-01  | S-02  | P-01  |       |       |       |       |       |       |

Do you have any suggestions on how you would make these changes? If this would be useful to you I'm happy to work together on a PR.

-----

I hacked in a quick support for multiple sample types by adding new logic after the following line:

https://github.com/dr01d3r/redcap-em-orca-specimen-tracking/blob/e05a7b0141df4d0de347a0b16b44765019d0e858/src/components/SpecimenForm.vue#L415 

```js
const specimenMatchesBox = (value, vm, model) => {
        // if box sample_type is mixed or mx, allow any value for the group sample_type in the specimen name but check the rest
        if (model.box_info.sample_type === 'mx') {
            let specimen_match = value.match(model.config.specimen_name_regex);
            // ignore if it's not a match - either empty or base regex validation failed
            if (specimen_match === null) return true;
            let is_valid = true;
            for (const [key, value] of Object.entries(model.box_info.box_name_parsed)) {
                let a, b;
                switch(key) {
                    case "visit":
                    case "aliquot_number":
                        a = parseInt(specimen_match.groups[key]);
                        b = parseInt(value);
                        // ensure they both parsed as integer, then compare
                        // a value of '00' allows any values
                        is_valid = is_valid && !isNaN(a) && !isNaN(b) && (b === 0 || a === b);
                        break;
                    default:
                        // ignore groups not specifically targeted for this validation
                        break;
                }
            }
            return is_valid;
        }
```

This allows me to enter arbitrary sample types to a mixed box.   
However, I have noticed that aliquot number appears to correspond to box position. In our case aliquot number resets with new sample types despite being associated with the same patient visit eg: 

- `ptid-v01-BC-01`
- `ptid-v01-BC-02`
- ...
- `ptid-v01-BC-0n`
- `ptid-v01-PG-01`
- `ptid-v01-PG-02`

Trying to enter `ptid-v01-PG-01` would result in an error because it tries to place it in a spot where `ptid-v01-BC-01` is.   
Here is a concrete example of the issue I'm having: 

![image](https://github.com/dr01d3r/redcap-em-orca-specimen-tracking/assets/32958525/41336f11-71d6-4315-930e-96287f77229c)

I realise that this sample schema is likely specific to our institution but some advice here would be appreciated.

**Comment by dr01d3r** [comment]
Hi @saipavan10-git thanks for the submission.

This is definitely something we're looking into, and are about to start a development cycle on this EM to support this, among other things.

The reason the box position is taken in this example is due to the `box_type` being used.  Our lab's "mixed" `box_type` (`00`) uses a very specific assignment algorithm to place specimens in the wells/positions.  Specimens are organized by participant, visit, aliquot number, given the configured number of visits and aliquots per visit.  So given these constraints, multiple sample types would not be feasible.

I would have to look a bit closer, but you might be able to achieve what you're after by using another box type, given your suggested modifications above, or something similar.

**Comment by saipavan10-git** [comment]
Thank you for your response. I understand that the "mixed" `box_type` does not support the functionality we are seeking. Is there another `box_type` that could accommodate this functionality or something similar?

If there isn't a suitable `box_type`, would it be necessary for us to create a new `box_type` to implement this functionality? If so, could you provide some guidance on how to go about this? Alternatively, are there specific areas in the existing code where we could make changes to accommodate this functionality?

Any guidance or resources you could provide would be greatly appreciated.

**Comment by dr01d3r** [comment]
Resolved by #4

@saipavan10-git @pbchase I know this is a bit delayed, but a new version of this module has just been submitted to the repo and should hopefully be approved and ready for use in the morning (11/19).  This v2 is a major overhaul of the functionality and configuration capabilities and should hopefully be much more able to meet your needs.

There is no longer a concept of "mixed box", but rather, you have 100% of the control now, through configurable box and specimen nomenclatures.  If you define a named capture group in the box name nomenclature, and the same named capture group in the sample name nomenclature, they must always match.  In the case of a mixed box, you simply omit a sample type from the box name and you're good to go!

We're still working on building out the documentation a bit more, but feel free to join our webinar tomorrow afternoon that discusses the module as well as a live demo and Q&A.

https://redcap.vumc.org/community/post.php?id=242943
